go/appencryption/session_cache_test.go

```diff
@@ -0,0 +1,379 @@
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+type closeSpy struct {
+	isClosed bool
+	mu       sync.Mutex
+}
+
+func (s *closeSpy) IsClosed() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.isClosed
+}
+
+func (s *closeSpy) SetClosed() {
+	s.mu.Lock()
+	s.isClosed = true
+	s.mu.Unlock()
+}
+
+type sessionBucket struct {
+	closeSpies map[*Session]*closeSpy
+	mu         sync.Mutex
+}
+
+func (b *sessionBucket) load(_ string) (*Session, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b.newSession(), nil
+}
+
+func (b *sessionBucket) newSession() *Session {
+	spy := &closeSpy{}
+
+	s := newSessionWithMockEncryption(func(_ mock.Arguments) {
+		b.mu.Lock()
+		defer b.mu.Unlock()
+
+		spy.SetClosed()
+	})
+	b.closeSpies[s] = spy
+
+	return s
+}
+
+func (b *sessionBucket) IsClosed(s *Session) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if spy, ok := b.closeSpies[s]; ok {
+		return spy.IsClosed()
+	}
+
+	return false
+}
+
+func newSessionBucket() *sessionBucket {
+	return &sessionBucket{
+		closeSpies: make(map[*Session]*closeSpy),
+	}
+}
+
+func newSessionWithMockEncryption(callbacks ...func(mock.Arguments)) *Session {
+	s := new(Session)
+	m := new(MockEncryption)
+	call := m.On("Close").Return(nil)
+
+	for i := range callbacks {
+		call = call.Run(callbacks[i])
+	}
+
+	sessionInjectEncryption(s, m)
+
+	return s
+}
+
+func TestNewSessionCache(t *testing.T) {
+	withEachEngine(t, func(t *testing.T, policy *CryptoPolicy) {
+		loader := func(id string) (*Session, error) {
+			return &Session{}, nil
+		}
+
+		cache := newSessionCache(loader, policy)
+		defer cache.Close()
+
+		require.NotNil(t, cache)
+	})
+}
+
+func withEachEngine(t *testing.T, testFunc func(*testing.T, *CryptoPolicy)) {
+	var engines = [...]string{
+		"mango",
+		"ristretto",
+	}
+
+	for i := range engines {
+		engine := engines[i]
+		policy := NewCryptoPolicy(
+			WithSessionCacheEngine(engine),
+		)
+
+		t.Run(fmt.Sprintf("WithEngine %s", engine), func(t *testing.T) {
+			testFunc(t, policy)
+		})
+	}
+}
+
+func TestNewSessionCachePanicsWithUnknownEngine(t *testing.T) {
+	loader := func(id string) (*Session, error) {
+		return &Session{}, nil
+	}
+
+	engine := "bogus"
+	policy := NewCryptoPolicy(
+		WithSessionCacheEngine(engine),
+	)
+
+	assert.PanicsWithValue(t, "invalid session cache engine: "+engine, func() {
+		newSessionCache(loader, policy)
+	})
+}
+
+func TestSessionCacheGetUsesLoader(t *testing.T) {
+	withEachEngine(t, func(t *testing.T, policy *CryptoPolicy) {
+		session := newSessionWithMockEncryption()
+
+		loader := func(id string) (*Session, error) {
+			return session, nil
+		}
+
+		cache := newSessionCache(loader, policy)
+		require.NotNil(t, cache)
+
+		defer cache.Close()
+
+		val, err := cache.Get("some-id")
+		require.NoError(t, err)
+		assert.Same(t, session, val)
+	})
+}
+
+func TestSessionCacheGetDoesNotUseLoaderOnHit(t *testing.T) {
```

Comment:
This test verifies that if we find a session in the cache, then the cache loader isn't called. Should/can we also have a check that verifies that when we find a session in the cache, then there are no metastore calls?